ARROW-15091: [C++][Doc] Document nodes in C++ streaming execution engine

vibhatha opened the pull request:
This PR's objective is to create a detailed documentation on using ExecPlan and available options. The task is divided into two sub-components. The first component is to work on a set of examples showing how to use these options with sample data. 

## Examples

- [x] Source-Sink
- [x] Scan-Sink
- [x] Scan-Filter-Sink
- [x] Scan-Project-Sink
- [x] Source-Aggregate-Sink
- [x] Scan-ConsumingSink
- [x] Scan-OrderBySink
- [x] Scan-HashJoin
- [x] Scan-SelectSink
- [x] Scan-Filter-Write
- [x] Scan-Union-Sink

## Documentation

- [x] General Description 
- [x] Option Description with code snippets

## Notes 

Some of the util functions were referred from the `test_utils`. 
There is a potential duplication of code due to that. Need 
feedback on best way to handle this.

lidavidm commented on cpp/examples/arrow/exec_plan_examples.cc:
```diff
@@ -0,0 +1,1122 @@
+#include <memory>
+#include <utility>
+
+#include "arrow/compute/api.h"
```
In general, try to be consistent with include style. Within Arrow code we use "" includes, but examples are "external" and generally use <> includes.

vibhatha replied:
Updating code.

lidavidm commented on cpp/examples/arrow/exec_plan_examples.cc:
```diff
@@ -0,0 +1,1122 @@
+
+#define SEP_STR "******"
+
+#define PRINT_BLOCK(msg)                                                     \
```
These could be functions and/or constants (`constexpr char kSep[] = "..."`)

vibhatha replied:
Updated

lidavidm commented on cpp/examples/arrow/exec_plan_examples.cc:
```diff
@@ -0,0 +1,1122 @@
+    const std::shared_ptr<arrow::DataType>& type, arrow::util::string_view json) {
+  std::shared_ptr<arrow::Array> out;
+
+  ABORT_ON_FAILURE(arrow::ipc::internal::json::ArrayFromJSON(type, json, &out));
```
I'm not sure we want to use the internal JSON format, both because it's not documented and because its format differs from the JSONL format the actual JSON reader uses. Can we use either the actual JSON reader or the CSV reader?

vibhatha replied:
Yes, we can do that. Should we hard code some values or use a real world dataset for this?

lidavidm replied:
I think we can keep hardcoding values just to keep it self-contained.

vibhatha replied:
Fixed with a csv based reader using in-memory string data.

lidavidm commented on cpp/examples/arrow/exec_plan_examples.cc:
```diff
@@ -0,0 +1,1122 @@
+  std::shared_ptr<arrow::Array> struct_array = GetArrayFromJSON(struct_type, json);
+
+  // Convert StructArray to RecordBatch
+  return *arrow::RecordBatch::FromStructArray(struct_array);
```
FWIW, there is a RecordBatchFromJSON.

lidavidm replied:
You can also directly create a RecordBatch from an ArrayVector (vector<shared_ptr<Array>>) instead of jumping through StructArray.

vibhatha replied:
This is WIP function as same as some other generators. Extracted some from internals. Will replace this with better functions.

vibhatha replied:
removed.

vibhatha replied:
Removed.

lidavidm commented on cpp/examples/arrow/exec_plan_examples.cc:
```diff
@@ -0,0 +1,1122 @@
+  for (const std::string& batch_json : json) {
+    batches.push_back(GetRecordBatchFromJSON(schema, batch_json));
+  }
+  return *arrow::Table::FromRecordBatches(schema, std::move(batches));
```
Similarly, there is a TableFromJSON.

vibhatha replied:
Removed.

lidavidm commented on cpp/examples/arrow/exec_plan_examples.cc:
```diff
@@ -0,0 +1,1122 @@
+                                ::arrow::internal::GetCpuThreadPool());
+
+    // ensure arrow::dataset node factories are in the registry
+    arrow::dataset::internal::Initialize();
```
Does this not get called automatically? User code should never have to call this.

lidavidm commented on cpp/examples/arrow/exec_plan_examples.cc:
```diff
@@ -0,0 +1,1122 @@
+}
+
+
+arrow::Status scan_sink_node_example() {
```
nit, but usually we use UpperCamelCase for function names ("Consume", "ScanSinkNodeExample"). We use snake_case only for const getters.

vibhatha replied:
Updated Naming convention

lidavidm commented on cpp/examples/arrow/exec_plan_examples.cc:
```diff
@@ -0,0 +1,1122 @@
+  std::vector<cp::ExecBatch> batches;
+  std::shared_ptr<arrow::Schema> schema;
+
+  arrow::AsyncGenerator<arrow::util::optional<cp::ExecBatch>> gen(bool parallel) const {
```
Would it make more sense to demonstrate InMemoryDataset here? This is using a lot of semi-internal machinery that I think will be confusing.

vibhatha replied:
One question related to this, the `SourceNodeOptions` requires the input

```c++
SourceNodeOptions(std::shared_ptr<Schema> output_schema,
                    std::function<Future<util::optional<ExecBatch>>()> generator) {...}
```

Are you referring to the return type of this function or about the internal components used in the function body.

lidavidm replied:
Ah, right. A `Dataset` can be turned into a `Scanner` and the `Scanner` can _internally_ build an ExecPlan, but that isn't exposed. However, that means we're stuck using internal APIs in an example which we should avoid as well.

vibhatha replied:
What is the best option for this? Not quite sure how to wrap this function?

lidavidm replied:
Sorry, I had thought there was a public conversion from Dataset/Scanner to ExecNode…it seems not. I pinged Weston below. However here I would want to make at least two changes:

1. Make it clear that this implementation is for ease of demonstration, not performance
2. Get rid of the `parallel` parameter and only demonstrate the non-parallel path (since it's an example anyways)

vibhatha replied:
Removed `parallel` option.

vibhatha replied:
> Can we add a placeholder section and link to that perhaps?
> […](#)
> On Tue, Jan 4, 2022, at 11:48, Vibhatha Lakmal Abeykoon wrote: ***@***.**** commented on this pull request. In docs/source/cpp/streaming_execution.rst <[#12033 (comment)](https://github.com/apache/arrow/pull/12033#discussion_r778232832)>: > + // data schema + auto schema = arrow::schema({...}) + + // source node options + auto source_node_options = arrow::compute::SourceNodeOptions{schema, gen}; + + // create a source node + ARROW_ASSIGN_OR_RAISE(arrow::compute::ExecNode * source, + arrow::compute::MakeExecNode("source", plan.get(), {}, + source_node_options)); + +``FilterNode`` +-------------- + +:class:`FilterNode`, as the name suggests, provide a container to define a data filtering criteria. +Filter can be written using :class:`arrow::compute::Expression`. For instance if the row values compute.rst contains a some of the function definitions, but not a detailed examples on how to use expressions. — Reply to this email directly, view it on GitHub <[#12033 (comment)](https://github.com/apache/arrow/pull/12033#discussion_r778232832)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AACQB37JP4IFUAYPITMT3JLUUMQFPANCNFSM5KUSEHEA>. Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. You are receiving this because you were mentioned.Message ID: ***@***.***>

@lidavidm should we create a new RST or just a sub-section within the same document?

lidavidm replied:
Hmm, it probably goes in its own section of compute.rst. That said, it occurs to me now that unless we intentionally fill out the section before 7.0.0, it would be rather confusing, and linking to the API docs for now (or just not linking if there's nothing to link to…?) is better.

vibhatha replied:
I can work on a separate PR for it and may be once it is reviewed properly we can link to this. 
I guess it won't take too much time. @westonpace @lidavidm WDYT?

westonpace replied:
I think I mentioned this below but the "scan" node is the correct way IMO to go from dataset + scan options -> exec node.  However, "ScanOptions" is something of a hidden class (especially in python) and there is a JIRA to hide it completely, exposing only ScannerBuilder so that is a bit of a conundrum.  Maybe ScannerBuilder could have a BuildExecNode method which creates a scan node.  @lidavidm thoughts?

westonpace replied:
For now I think it is fine to document everything but scan and we can add scan later in a follow-up PR once we align on this.

lidavidm commented on cpp/examples/arrow/exec_plan_examples.cc:
```diff
@@ -0,0 +1,1122 @@
+}
+
+arrow::Status source_sink_example() {
+  cp::ExecContext exec_context(arrow::default_memory_pool(),
```
Hmm, are these arguments not the defaults?

vibhatha replied:
Yes, you're correct. Will remove this.

lidavidm commented on cpp/examples/arrow/exec_plan_examples.cc:
```diff
@@ -0,0 +1,1122 @@
+  return out;
+}
+
+std::shared_ptr<arrow::internal::ThreadPool> MakeIOThreadPool() {
+  auto maybe_pool = arrow::internal::ThreadPool::MakeEternal(/*threads=*/8);
+  if (!maybe_pool.ok()) {
+    maybe_pool.status().Abort("Failed to create global IO thread pool");
+  }
+  return *std::move(maybe_pool);
+}
+
+arrow::internal::ThreadPool* GetIOThreadPool() {
+  static std::shared_ptr<arrow::internal::ThreadPool> pool = MakeIOThreadPool();
+  return pool.get();
+}
```
This seems unused.

vibhatha replied:
Need to clean up some of these functions

vibhatha replied:
Removed.

lidavidm commented on cpp/examples/arrow/exec_plan_examples.cc:
```diff
@@ -0,0 +1,1122 @@
+    arrow::AsyncGenerator<arrow::util::optional<cp::ExecBatch>> sink_gen;
+
+    std::string root_path = "";
+    std::string uri = "file:///Users/vibhatha/sandbox/test";
```
Remember to un-hardcode this :) 

We can link to gflags for command line argument parsing if necessary.

vibhatha replied:
Removed hardcoded values and added cmd args.

lidavidm commented:
BTW, to get CI to run,  the [WIP] needs to be removed from the issue title (and the branch needs to have a new commit/be rebased)

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -175,9 +175,607 @@ their completion::
+Constructing ``ExecNode`` using Options
+=======================================
+
+Using the execution plan we can construct varioud execution queries. 
```
```suggestion
Using the execution plan we can construct various queries. 
```

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -175,9 +175,607 @@ their completion::
+=======================================
+
+Using the execution plan we can construct varioud execution queries. 
+To construct such queries, we have provided a set of containers or 
```
containers -> maybe "building blocks"?

vibhatha replied:
Yeah that sounds better.

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -175,9 +175,607 @@ their completion::
+5. :class:`SinkNode`
+6. :class:`ConsumingSinkNode`
+7. :struct:`OrderBySinkNode`
+8. SelectK-SinkNode
```
If some are undocumented, we should add the documentation so we can link them.

That said, I'm not sure how useful the API docs for nodes are, it might even be confusing.

vibhatha replied:
May be we can say about node operation without saying the node names. For instance join operation, filter operation, etc. I think it’s important to expose the available operations ? WDYT?

lidavidm replied:
I think we should just refer to them by their factory names, `"scan"`, `"filter"`, etc. As said below I don't think it really helps to link to the API docs for a node class.

vibhatha replied:
Yes, this would simplify and I can also avoid mentioning the nodes and rather talk about the operation and its options.

vibhatha replied:
Nice suggestion, table is neater and link to additional docs. I will update the docs.

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -175,9 +175,607 @@ their completion::
+:struct:`arrow::compute::SourceNode` can be considered as an entry point to create a streaming execution plan. 
+A source node can be constructed as follows.
+
+:class:`arrow::compute::SoureNodeOptions` are used to create the :struct:`arrow::compute::SourceNode`. 
```
```suggestion
:class:`arrow::compute::SourceNodeOptions` are used to create the :struct:`arrow::compute::SourceNode`. 
```

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -175,9 +175,607 @@ their completion::
     Future<> complete = plan->finished();
 
+Constructing ``ExecNode`` using Options
+=======================================
```
It feels like this should all go _after_ the docs about constructing an ExecPlan below, since that's where the actual code to create an exec node is introduced.

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -175,9 +175,607 @@ their completion::
+`std::function<arrow::Future<arrow::util::optional<arrow::compute::ExecBatch>>()>` 
+are required to create this option::
+
+    // data generator
+    arrow::AsyncGenerator<arrow::util::optional<cp::ExecBatch>> gen() { ... }
+    
+    // data schema 
+    auto schema = arrow::schema({...})
+    
+    // source node options
+    auto source_node_options = arrow::compute::SourceNodeOptions{schema, gen};
+    
+    // create a source node
+    ARROW_ASSIGN_OR_RAISE(arrow::compute::ExecNode * source,
+                            arrow::compute::MakeExecNode("source", plan.get(), {}, 
+                            source_node_options));
```
This isn't valid C++ or clear pesudocode (missing semicolons, also the declaration of `gen`)

lidavidm replied:
The _overall_ thing missing is probably a way to get an `AsyncGenerator<optional<ExecBatch>>` from a `Scanner`, as noted above.

vibhatha replied:
@lidavidm the declation for gen can be included, but we are not 100% sure whether to include the iternal APIs.
I am waitng for that to decide the best way to point out this.

vibhatha replied:
@lidavidm this thread is not yet in the conversation page.

lidavidm replied:
You have to expand the hidden items in the conversation view. Unsubmitted comments don't show up at all for other people, as far as I'm aware.

lidavidm commented on cpp/examples/arrow/execution_plan_documentation_examples.cc:
```diff
@@ -0,0 +1,1160 @@
+  std::vector<cp::ExecBatch> batches;
+  std::shared_ptr<arrow::Schema> schema;
+
+  arrow::AsyncGenerator<arrow::util::optional<cp::ExecBatch>> gen(bool parallel) const {
+    auto opt_batches = ::arrow::internal::MapVector(
+        [](cp::ExecBatch batch) { return arrow::util::make_optional(std::move(batch)); },
+        batches);
+
+    arrow::AsyncGenerator<arrow::util::optional<cp::ExecBatch>> gen;
+
+    if (parallel) {
+      // emulate batches completing initial decode-after-scan on a cpu thread
+      gen = arrow::MakeBackgroundGenerator(
+                arrow::MakeVectorIterator(std::move(opt_batches)),
+                ::arrow::internal::GetCpuThreadPool())
+                .ValueOrDie();
+
+      // ensure that callbacks are not executed immediately on a background thread
+      gen = arrow::MakeTransferredGenerator(std::move(gen),
+                                            ::arrow::internal::GetCpuThreadPool());
+    } else {
+      gen = arrow::MakeVectorGenerator(std::move(opt_batches));
+    }
+
+    return gen;
+  }
```
Probably we should take `Scanner::ScanBatchesUnorderedAsync` and use `arrow::MakeMapGenerator` to convert it into a generator of ExecBatch, instead of using internal utilities here and making confusing comments about threads (which users should not need to deal with).

lidavidm replied:
Either that, or `SourceNodeOptions` should implement a `MakeFromScanner`. Thoughts cc @westonpace?

lidavidm replied:
Oh, seems like I already pinged you here. @vibhatha this is really the decision. I think adding one API would fix all the comments about `gen` and `internal::Initialize`.

vibhatha replied:
cleaned up the function and removed parallel option

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -175,9 +175,607 @@ their completion::
+``FilterNode``
+--------------
+
+:class:`FilterNode`, as the name suggests, provide a container to define a data filtering criteria. 
```
I think we can be more concise here. Example

lidavidm replied:
```suggestion
:class:`FilterNode`, as the name suggests, keeps only rows matching a given expression. 
```

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -175,9 +175,607 @@ their completion::
+container to define the aggregation criterion. These options can be 
+selected from `arrow::compute` options. 
+
+1. `ScalarAggregateOptions`
```
We shouldn't talk about individual options classes here. We should link to the relevant documentation instead. Most of this should actually be in the API docs, and if not, should be moved to documentation comments in the source code.

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -175,9 +175,607 @@ their completion::
+--------------
+
+:class:`FilterNode`, as the name suggests, provide a container to define a data filtering criteria. 
+Filter can be written using :class:`arrow::compute::Expression`. For instance if the row values
+of a particular column needs to be filtered by a boundary value, ex: all values of column b
+greater than 3, can be written using :class:`arrow::compute::FilterNodeOptions` as follows::
```
```suggestion
Filters can be written using :class:`arrow::compute::Expression`. For example, if we wish
to keep rows of column ``b`` greater than 3, then we can use the following expression::
```

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -175,9 +175,607 @@ their completion::
+                            // previous input
+                            {scan}, 
+                            //filter node options
+                            arrow::compute::FilterNodeOptions{filter_opt}));
```
The code example here could be formatted (e.g. there's random indentation on the macro, and missing indentation for the MakeExecNode call)

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -175,9 +175,607 @@ their completion::
+      ARROW_ASSIGN_OR_RAISE(filter, arrow::compute::MakeExecNode("filter", 
+                            // plan
+                            plan.get(),
+                            // previous input
```
```suggestion
                            // input node
```

lidavidm replied:
for clarity?

vibhatha replied:
sure. make sense.

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -175,9 +175,607 @@ their completion::
+:class:`arrow::compute::ProjectNodeOptions` component which requires, 
+a :class:`arrow::compute::Expression`, names of the project columns (names are not provided, 
+the string representations of exprs will be used) and a boolean flag to determine 
+synchronous/asynchronous nature (by default asynchronous option is set to `true`). 
```
async_mode should be treated as internal and we shouldn't talk about it.

lidavidm replied:
(It's likely to go away entirely sooner or later.)

vibhatha replied:
Sure, will remove that.

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -175,9 +175,607 @@ their completion::
+:class:`ProjectNode` executes expressions on input batches and produces new batches. 
+Each expression will be evaluated against each batch which is pushed to this 
+node to produce a corresponding output column. This is exposed via 
+:class:`arrow::compute::ProjectNodeOptions` component which requires, 
```
component -> class? 

We should call a spade a spade and avoid overly vague terminology. Or we can just say, "ProjectNode is created from ProjectNodeOptions which requires..."

vibhatha replied:
Sure, I understand.

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -175,9 +175,607 @@ their completion::
+Each expression will be evaluated against each batch which is pushed to this 
+node to produce a corresponding output column. This is exposed via 
+:class:`arrow::compute::ProjectNodeOptions` component which requires, 
+a :class:`arrow::compute::Expression`, names of the project columns (names are not provided, 
```
nit: "_if_ names are not provided…"

and it takes a vector of expressions, no?

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -175,9 +175,607 @@ their completion::
+--------------
+
+:class:`FilterNode`, as the name suggests, provide a container to define a data filtering criteria. 
+Filter can be written using :class:`arrow::compute::Expression`. For instance if the row values
```
Is there prose documentation on Expression that we can link to? Or a section in the API docs that lists all the helpers like `greater`, `call`, etc.?

vibhatha replied:
Yeah, I see your point. I was also not sure about some other parts I added like this. 
Better to link to a doc if exist or I can create an issue to make one. Let me work on this.

vibhatha replied:
compute.rst contains a some of the function definitions, but not a detailed examples on how to use expressions.

vibhatha replied:
@lidavidm should we add this to a separate PR? (PR including compute.rst and links to this doc?)

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -175,9 +175,607 @@ their completion::
+
+Creating a project node::
+
+    arrow::compute::ExecNode *project;
+        ARROW_ASSIGN_OR_RAISE(project, 
+            arrow::compute::MakeExecNode("project", 
+            // plan
+            plan.get(),
+            // previous node 
+            {scan},
+            // project node options 
+            arrow::compute::ProjectNodeOptions{{a_times_2}}));
```
Same comments about the formatting here.

lidavidm replied:
Also, the Arrow style guide would declare `arrow::compute::ExecNode* project;`.

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -175,9 +175,607 @@ their completion::
+            // project node options 
+            arrow::compute::ProjectNodeOptions{{a_times_2}}));
+
+``ScalarAggregateNode``
```
General comment: each of these sections can be made more consistent, in terms of wording and structure.

> (node name) does (X). It is constructed from (options class), which requires (parameters). (More description of parameters if necessary)
> 
> For example, to create a (node name) that does (specific example)::
>     (code example follows)

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -175,9 +175,607 @@ their completion::
+``ScalarAggregateNode``
+-----------------------
+
+:class:`ScalarAggregateNode` is an :class:`ExecNode` which provides various 
+aggregation options. The :class:`arrow::compute::AggregateNodeOptions` provides the 
+container to define the aggregation criterion. These options can be 
+selected from `arrow::compute` options. 
```
This is rather vague. If we're explaining what a filter is, then we need to explain what an aggregation is. Or vice versa, if we _don't_ explain what an aggregation is, then we shouldn't be explaining a filter, either.

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -175,9 +175,607 @@ their completion::
+                              aggregate_options));
+
+
+Scan-Node
```
```suggestion
``ScanNode``
```

vibhatha replied:
But there is no entity like that in the source code, this was another unclear part. 
The definition set is not consistent. Some of them have a node and some have the 
functionality but a node is not exposed even internally. For instance "MakeScaneNode" 
is not a function exposed.

lidavidm replied:
Yup, I see your point. I think we should just refer to these as "scan", "filter", "project" etc. since users should never need to interact with the actual classes.

vibhatha replied:
Replaced the nodes with `scan`, `project`, etc operations and didn't explicitly discussed about nodes.

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -175,9 +175,607 @@ their completion::
+Scan-Node
+---------
+
+There is no class or struct defined as ScanNode in the source. 
```
I think this means we should probably not refer to individual node classes, but only to the general node names, since we only expose them via the factory interface. Then the scan node won't be an odd exception.

lidavidm replied:
Also, again, because the API docs for individual node classes are useless at best, and potentially misleading at worst.

vibhatha replied:
I understand. I can modify this into options rather than nodes.

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -175,9 +175,607 @@ their completion::
+
+There is no class or struct defined as ScanNode in the source. 
+But :class:`arrow::compute::ScanNodeOptions` container includes the options
+passed to `MakeScanNode` internal function which creates an :class:`ExecNode`
```
We don't need to talk about implementation, only purpose. And actually, we haven't talked about what a scan node even does.

lidavidm replied:
We should also link to the general datasets documentation, and perhaps directly to the section about scanning.

vibhatha replied:
linked.

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -175,9 +175,607 @@ their completion::
+`std::shared_ptr<arrow::compute::ScanOptions>`, 
+`std::shared_ptr<arrow::util::AsyncToggle>`. 
+
+The :class:`arrow::compute::ScanOptions` includes the scaning options::
```
The code block here is rather confusing. Also, we haven't explained what the parameters are.

westonpace replied:
If we link to the documentation on dataset scanning then we don't need to do a lot of explaining around what `ScanOptions` are.  Just assume the user is starting with a dataset and some scan options.  I also think we can avoid any examples of augmented fields.

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -175,9 +175,607 @@ their completion::
+defined in the :class:`arrow::compute::ScanNodeOptions` and this requires, 
+`std::shared_ptr<arrow::dataset::Dataset>`, 
+`std::shared_ptr<arrow::compute::ScanOptions>`, 
+`std::shared_ptr<arrow::util::AsyncToggle>`. 
```
Hmm. Do we even want to document the scan node, given it has this internal parameter that users can't really provide?

lidavidm replied:
I guess in general, now I'm questioning the goals of what we're documenting here, since a lot of this is inaccessible or confusing to users. And if it's meant as developer documentation, I wonder if there's a better place for it, or if the focus shouldn't be on the documentation comments themselves.

vibhatha replied:
I see your point. May be I can move some of these stuff to a notion page for internal reference. 
And keep a neater high level details for the user. This documentation is meant to be a user doc.

lidavidm replied:
We shouldn't keep documentation internal just because it is user facing (we may work for the same place, but we should remember that this is an open-source project with contributors from different organizations). I think if there's anything useful to say about these options, they should be in doc comments in the source.

westonpace replied:
`AsyncToggle` is optional and it does not need to be provided (and now I'm wondering why it is part of scan node and not source node).  I do think we want to document scan node (just leave that option out).

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -175,9 +175,607 @@ their completion::
+:class:`SinkNode` can be considered as the output or final node of an streaming 
+execution definition. :class:`arrow::compute::SinkNodeOptions` interface is used to pass 
+the required options. Requires 
+`std::function<arrow::Future<arrow::util::optional<arrow::compute::ExecBatch>>()>* generator`
```
Note that this has a user-friendly alias, `function<Future<T>()>` can be written as `arrow::AsyncGenerator<T>`

westonpace replied:
Ah, but I don't know if `AsyncGenerator` is an external alias.  I could go either way.

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -175,9 +175,607 @@ their completion::
+will not finish until the consumption is finished.  Use SinkNode if you are
+transferring the ownership of the data to another system.  
+Use :class:`arrow::compute::ConsumingSinkNode` if the data is being consumed within the exec 
+plan (i.e. the exec plan should not complete until the consumption has completed).
```
This is nearly tautological. I think the general purpose is that `ConsumingSinkNode` is used to consume the output within the exec plan, instead of getting a stream of batches to read.

vibhatha replied:
Yes, I missed this point. I went through the code and you're correct.

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -175,9 +175,607 @@ their completion::
+``OrderBySinkNode``
+-------------------
+
+This is an extension to the :class:`SinkNode` definition and provides the ability
+to guarantee the ordering of the stream by providing the,
+:class:`arrow::compute::OrderBySinkNodeOptions`. 
+Here the :class:`arrow::compute::SortOptions` are provided to define which columns 
+are used for sorting and under which criterion.
```
```suggestion
This is variant of :class:`SinkNode` that sorts the results according to
:class:`arrow::compute::OrderBySinkNodeOptions`. 
Here :class:`arrow::compute::SortOptions` are provided to define which columns 
are used for sorting and whether to sort by ascending or descending values.
```

vibhatha replied:
From the description, omitted the node related descriptions, just made the connection via the operation type.

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -175,9 +175,607 @@ their completion::
+        plan.get(), {source},
+        arrow::compute::SelectKSinkNodeOptions{options, &sink_gen}));
+
+Scan-Node
```
Didn't we talk about scan node above?

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -175,9 +175,607 @@ their completion::
+Option definitions for :class:`arrow::dataset::ScanNodeOptions`:: 
+
+
+    /// A row filter (which will be pushed down to partitioning/reading if supported).
```
This is formatted oddly, plus, this all comes from ScanOptions, no?

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -175,9 +175,607 @@ their completion::
+The option to write a result to a file format is provided by this execution node type. 
+A definition doesn't exist as an :class:`ExecNode`, but the write options are provided
+via the :class:`arrow::dataset::WriteNodeOptions` and defined using 
+:class::`arrow::dataset::FileSystemDatasetWriteOptions`, `std::shared_ptr<arrow::Schema>`,
```
We need double backticks here.

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -175,9 +175,607 @@ their completion::
+Write-Node
+----------
+
+The option to write a result to a file format is provided by this execution node type. 
```
> The "write" node writes the results of a query into files on disk. Options for file format (Feather, CSV, Parquet, etc.), partitioning scheme, and so on are provided.

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -175,9 +175,607 @@ their completion::
+-------------
+
+:class:`UnionNode` is the :class:`ExecNode` interface to perform a union 
+operation on two datasets. The union operation can be executed
```
This can be confusing given the various uses of union in SQL. Specifically, UnionNode combines two streams of data with the same schema into one, and does not modify or look at any of the data itself.

lidavidm commented on cpp/examples/arrow/execution_plan_documentation_examples.cc:
```diff
@@ -0,0 +1,1160 @@
+
+constexpr char kSep[] = "******";
+
+#define PRINT_BLOCK(msg)                                                     \
```
Do we really need to use macros for printing? Why not a function?

vibhatha replied:
No we can remove it and just use std::cout

lidavidm commented on cpp/examples/arrow/execution_plan_documentation_examples.cc:
```diff
@@ -0,0 +1,1160 @@
+        exec_context.memory_pool());
+
+    // // // start the ExecPlan
+    ABORT_ON_FAILURE(plan->StartProducing());
```
Why abort here instead of `ARROW_RETURN_NOT_OK`?

vibhatha replied:
Yes, it suits better. Replaced.

lidavidm commented on cpp/examples/arrow/execution_plan_documentation_examples.cc:
```diff
@@ -0,0 +1,1160 @@
+  switch (mode) {
+    case SOURCE_SINK:
+      PRINT_BLOCK("Source Sink Example");
+      CHECK_AND_CONTINUE(SourceSinkExample())
```
We can mostly avoid defining new macros if we instead refactor this into a function that returns Status, and then just check the status once in `main`.

lidavidm commented on cpp/examples/arrow/execution_plan_documentation_examples.cc:
```diff
@@ -0,0 +1,1160 @@
+std::shared_ptr<arrow::Array> GetArrayFromJSON(
+    const std::shared_ptr<arrow::DataType>& type, arrow::util::string_view json) {
+    std::shared_ptr<arrow::Array> out;
+    ABORT_ON_FAILURE(arrow::ipc::internal::json::ArrayFromJSON(type, json, &out));
```
Hmm, we really shouldn't use the internal JSON parser.

vibhatha replied:
Removed this function.

lidavidm commented on cpp/examples/arrow/execution_plan_documentation_examples.cc:
```diff
@@ -0,0 +1,1160 @@
+std::string GetDataAsCsvString() {
+    std::string data_str = "";
+
+    data_str.append("a,b\n");
```
We can use raw strings ([(6) here](https://en.cppreference.com/w/cpp/language/string_literal)).

lidavidm commented on cpp/examples/arrow/execution_plan_documentation_examples.cc:
```diff
@@ -0,0 +1,1160 @@
+}
+
+arrow::Status CreateDataSetFromCSVData(
+    std::shared_ptr<arrow::dataset::InMemoryDataset> &dataset) {
```
Use pointers for out parameters.

lidavidm replied:
Or really, just return `Result<std::shared_ptr<Dataset>>`.

vibhatha replied:
That’s better. Sure I will update.

lidavidm commented on cpp/examples/arrow/execution_plan_documentation_examples.cc:
```diff
@@ -0,0 +1,1160 @@
+
+std::shared_ptr<arrow::dataset::Dataset> CreateDataset() {
+    std::shared_ptr<arrow::dataset::InMemoryDataset> im_dataset;
+    ABORT_ON_FAILURE(CreateDataSetFromCSVData(im_dataset));
```
In general, let's consistently use `Result` and `Status` and friends, or consistently abort on each assignment, but let's not mix the two styles.

vibhatha replied:
Good point. 👍

vibhatha replied:
For the functions which should return some value I used results and for all the samples 
added the Status.

lidavidm commented on cpp/examples/arrow/execution_plan_documentation_examples.cc:
```diff
@@ -0,0 +1,1160 @@
+    return im_dataset;
+}
+
+arrow::Status exec_plan_end_to_end_sample() {
```
For consistency, use UpperCamelCase naming.

lidavidm replied:
For each sample, can we have a brief doc comment explaining what the purpose of the sample is?

vibhatha replied:
Yes that’s a nice idea. Will update each example.

lidavidm commented on cpp/examples/arrow/execution_plan_documentation_examples.cc:
```diff
@@ -0,0 +1,1160 @@
+                                ::arrow::internal::GetCpuThreadPool());
+
+    // ensure arrow::dataset node factories are in the registry
+    arrow::dataset::internal::Initialize();
```
We shouldn't need to call this, right?

lidavidm replied:
In general, if we're calling anything labeled `internal` in an example, then something else needs fixing, I think.

vibhatha replied:
Hmm, does this function call do some registration (AddFactory) of ops associated with the dataset namespace?

lidavidm replied:
It is called already when you use a relevant API. Do not call it directly, or if you need to call it directly, then it should be fixed. We should not use internal APIs in an example that is meant for users.

vibhatha replied:
Sure, will remove this.

vibhatha replied:
/// Register dataset-based exec nodes with the exec node registry
///
/// This function must be called before using dataset ExecNode factories
ARROW_DS_EXPORT void Initialize();

this comment is in plan.h

westonpace replied:
The challenge is that factory registration for most things happens in `arrow::compute::default_exec_factory_registry()` but the `compute` namespace should not know about the `dataset` namespace.

This is one of those things I think we can get rid of with an `engine` namespace (I created https://issues.apache.org/jira/browse/ARROW-15238 so I don't forget about this) but I don't think I'll get to that PR until after 7.0.0.

In the meantime I vote for simply not documenting but including in the examples so they compile and work.

lidavidm replied:
Ok, that seems reasonable. We can just do it once in main() then.

lidavidm commented on cpp/examples/arrow/execution_plan_documentation_examples.cc:
```diff
@@ -0,0 +1,1160 @@
+    // // plan mark finished
+    plan->finished().Wait();
+
+    return arrow::Status::OK();
```
We should return the status from `plan->finished()` to catch if the plan errors.

westonpace commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,601 @@ Datasets may be scanned multiple times; just make multiple scan
+``ProjectNode``
+---------------
+
+:class:`ProjectNode` executes expressions on input batches and produces new batches. 
+Each expression will be evaluated against each batch which is pushed to this 
+node to produce a corresponding output column. This is exposed via 
```
This wording is a little confusing to me.  Expressions do not create batches.  The output of the expressions is arrays.  Maybe something like:

```
:class:`ProjectNode` rearranges, deletes, transforms, and creates columns.
Each output column is computed by evaluating an expression
against the source record batch.  This is exposed via...
```

westonpace commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,601 @@ Datasets may be scanned multiple times; just make multiple scan
+Each expression will be evaluated against each batch which is pushed to this 
+node to produce a corresponding output column. This is exposed via 
+:class:`arrow::compute::ProjectNodeOptions` component which requires, 
+a :class:`arrow::compute::Expression`, names of the project columns (names are not provided, 
```
```suggestion
a :class:`arrow::compute::Expression`, names for the output columns (names are not provided, 
```

westonpace commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,601 @@ Datasets may be scanned multiple times; just make multiple scan
+
+1. `ScalarAggregateOptions`
+
+In this aggregation mode, using option, `skip_nulls` the null values are ignored.
```
"In this aggregation mode"...

What is an aggregation mode?

vibhatha replied:
removed these content and linked to existing aggregation docs.

westonpace commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,601 @@ Datasets may be scanned multiple times; just make multiple scan
+``SinkNode``
+------------
+
+:class:`SinkNode` can be considered as the output or final node of an streaming 
+execution definition. :class:`arrow::compute::SinkNodeOptions` interface is used to pass 
+the required options. Requires 
```
Might be worth mentioning that an exec plan should only have one "terminal" node.  I don't think we actually prevent this but "bad things may happen someday"

westonpace commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,601 @@ Datasets may be scanned multiple times; just make multiple scan
+                                                  arrow::compute::SinkNodeOptions{&sink_gen}));
+
+
+The output can be obtained as a table::
```
Not for this PR but I wonder if we might want to someday just create a `to_table` sink node which hides the complexity of aysnc generators.  The user could just use a scan node as input and a ToTableNode as output and never be aware of any of that.

vibhatha replied:
I think it is a nice idea. It will be very useful for the high level C++ users.

vibhatha replied:
I would like to work on it, if we are moving for such a functionality.

westonpace commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,601 @@ Datasets may be scanned multiple times; just make multiple scan
+SelectK-Node
+------------
+
+There is no Select-K-SinkNode available as an entity within the source, but the behavior 
```
Since the "node classes" themselves are really an implementation detail we don't need to call out the fact that a node doesn't actually exist or that we are reusing a node.

vibhatha replied:
Yes, @lidavidm also mentioned previously. The nodes will be removed and we will only document the 
options and mention them as operations instead of nodes, "scan", "source", etc.

westonpace commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,601 @@ Datasets may be scanned multiple times; just make multiple scan
+on multiple data sources(:class:`ExecNodes`).
+
+The following example demonstrates how this can be achieved using 
+two data sources. Following a union operations the output is obtained using 
+a aggregation operation. 
```
Why include the aggregation operation here?

vibhatha replied:
Replaced by referring to the existing docs.

westonpace replied:
I think I wasn't clear enough.  Why do you need an aggregate node in the example for union?  It seems you could just have Source -> Union -> Sink

westonpace reviewed (commented):
This is a great start and a lot of information.  Thanks for working on this.  I haven't looked through the examples in detail yet.  However, at a high level it seems like a large portion of the example is generating sample input.  I wonder if it might make more sense to just include some small binary input files in the example directory.

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,601 @@ Datasets may be scanned multiple times; just make multiple scan
+are required to create this option::
+
+    // data generator
+    arrow::AsyncGenerator<arrow::util::optional<cp::ExecBatch>> gen() { ... }
```
I think that's ok then, but it might be nice to explain that this represents a generator, since it may not otherwise be clear that you should expect to call the std::function repeatedly.

vibhatha commented on cpp/examples/arrow/execution_plan_documentation_examples.cc:
```diff
@@ -0,0 +1,1234 @@
+
+    cp::ExecNode* project;
+
+    ARROW_ASSIGN_OR_RAISE(project,
+                            cp::MakeExecNode("augmented_project", plan.get(), {filter},
+                                            cp::ProjectNodeOptions{{a_times_2}}));
```
"project" should be good for this example, also we can add a note on "augmented_project" ?

vibhatha commented on cpp/examples/arrow/execution_plan_documentation_examples.cc:
```diff
@@ -0,0 +1,1234 @@
+
+    cp::Expression a_times_2 = cp::call("multiply", {cp::field_ref("a"), cp::literal(2)});
+    options->projection =
+        cp::call("make_struct", {a_times_2}, cp::MakeStructOptions{{"a * 2"}});
```
I replaced with an empty materialisation. Is this okay?

westonpace replied:
Yes, although my hope would be that it doesn't need to be set at all.  Do you get an error if this is omitted?  We have the ScannerBuilder for creating ScanOptions when doing a normal scan which helps with setting the projection options.  I think we will want to revisit all of these scan options at some point after 7.0.0

vibhatha replied:
When this is omitted, I get an segfault. I didn't go deep into checking why? Should we analyze it for this PR?

vibhatha commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,601 @@ Datasets may be scanned multiple times; just make multiple scan
+greater than 3, can be written using :class:`arrow::compute::FilterNodeOptions` as follows::
+
+    // a > 3
+    arrow::compute::Expression filter_opt = arrow::compute::greater(
+                                  arrow::compute::field_ref("a"), 
+																	arrow::compute::literal(3));
```
I think I have added an additional tab or 2. Misunderstood that the tabs are required to highlight the code. 
I fixed it through out the rest of the code snippets.

lidavidm commented:
Can we add a placeholder section and link to that perhaps?

On Tue, Jan 4, 2022, at 11:48, Vibhatha Lakmal Abeykoon wrote:
> 
> 
> ***@***.**** commented on this pull request.
> 
> 
> In docs/source/cpp/streaming_execution.rst <https://github.com/apache/arrow/pull/12033#discussion_r778232832>:
> 
> > +    // data schema 
> +    auto schema = arrow::schema({...})
> +    
> +    // source node options
> +    auto source_node_options = arrow::compute::SourceNodeOptions{schema, gen};
> +    
> +    // create a source node
> +    ARROW_ASSIGN_OR_RAISE(arrow::compute::ExecNode * source,
> +                            arrow::compute::MakeExecNode("source", plan.get(), {}, 
> +                            source_node_options));
> +
> +``FilterNode``
> +--------------
> +
> +:class:`FilterNode`, as the name suggests, provide a container to define a data filtering criteria. 
> +Filter can be written using :class:`arrow::compute::Expression`. For instance if the row values
> 
> compute.rst contains a some of the function definitions, but not a detailed examples on how to use expressions.
> 
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/apache/arrow/pull/12033#discussion_r778232832>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AACQB37JP4IFUAYPITMT3JLUUMQFPANCNFSM5KUSEHEA>.
> Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. 
> You are receiving this because you were mentioned.Message ID: ***@***.***>
>

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,484 @@ Datasets may be scanned multiple times; just make multiple scan
+:class:`arrow::dataset::ScanNodeOptions`. This option contains a set of definitions. 
+The :ref:`dataset<cpp-dataset-reading>` API also use scanner for processing data.
+
+Option definitions for :class:`arrow::dataset::ScanNodeOptions`:: 
```
I still don't think we need/want to copy-paste the definition of ScanOptions into the source here…it will get out of sync right away.

lidavidm commented:
Also, just a thought about some of the longer code blocks in reST…how about using the `literalinclude` directive to reference the example instead? It will be a little painful to sync the line numbers with reST (you can somewhat avoid this by using the options of the directive) but that way the code will reflect what is actually in the example and will get formatted properly.

vibhatha commented:
> Also, just a thought about some of the longer code blocks in reST…how about using the `literalinclude` directive to reference the example instead? It will be a little painful to sync the line numbers with reST (you can somewhat avoid this by using the options of the directive) but that way the code will reflect what is actually in the example and will get formatted properly.

@lidavidm I added the code refereces as you suggested. It is neater. And in addition I also added the sub-example for each operation. It would be easier for the reader. And users can use the full `.cc` and read more.

westonpace commented:
I've been capturing some of the usability improvements we are identifying here: https://issues.apache.org/jira/browse/ARROW-15257

westonpace commented on cpp/examples/arrow/execution_plan_documentation_examples.cc:
```diff
@@ -0,0 +1,1125 @@
+  options->use_async = true;
+  // specify the filter
+  cp::Expression filter_opt = cp::greater(cp::field_ref("a"), cp::literal(3));
+  // set filter for scanner : on-disk / push-down filtering.
+  options->filter = filter_opt;
```
```suggestion
  // set filter for scanner : on-disk / push-down filtering.
  // This step can be skipped if you are not reading from disk.
  options->filter = filter_opt;
```

westonpace commented on cpp/examples/arrow/execution_plan_documentation_examples.cc:
```diff
@@ -0,0 +1,1125 @@
+  ARROW_ASSIGN_OR_RAISE(filter, cp::MakeExecNode("filter", plan.get(), {scan},
+                                                 cp::FilterNodeOptions{filter_opt}));
+
+  // // finally, pipe the project node into a sink node
```
```suggestion
  // // finally, pipe the filter node into a sink node
```

westonpace commented on cpp/examples/arrow/execution_plan_documentation_examples.cc:
```diff
@@ -0,0 +1,1125 @@
+  arrow::AsyncGenerator<arrow::util::optional<cp::ExecBatch>> sink_gen;
+  ARROW_ASSIGN_OR_RAISE(std::ignore, cp::MakeExecNode("sink", plan.get(), {filter},
+                                                      cp::SinkNodeOptions{&sink_gen}));
+  // // translate sink_gen (async) to sink_reader (sync)
+  std::shared_ptr<arrow::RecordBatchReader> sink_reader = cp::MakeGeneratorReader(
+      dataset->schema(), std::move(sink_gen), exec_context.memory_pool());
+
+  // // validate the ExecPlan
+  ABORT_ON_FAILURE(plan->Validate());
+  std::cout << "Exec Plan created " << plan->ToString() << std::endl;
+  // // start the ExecPlan
+  ARROW_RETURN_NOT_OK(plan->StartProducing());
+
+  // // collect sink_reader into a Table
+  std::shared_ptr<arrow::Table> response_table;
+  ARROW_ASSIGN_OR_RAISE(response_table,
+                        arrow::Table::FromRecordBatchReader(sink_reader.get()));
+
+  std::cout << "Results : " << response_table->ToString() << std::endl;
+  // // plan stop producing
+  plan->StopProducing();
+  // /// plan marked finished
+  auto futures = plan->finished();
+  ARROW_RETURN_NOT_OK(futures.status());
+  futures.Wait();
+
+  return arrow::Status::OK();
```
Since we repeat these steps a number of times (especially since these examples are not trying to demonstrate the sink node) maybe we can wrap them up in an `ExecutePlanAndCollectAsTable` method or something like that.  This way the example snippet can be focused on the node in question.

vibhatha replied:
Neat !!!. Created a separate method and refactored the code.

westonpace commented on cpp/examples/arrow/execution_plan_documentation_examples.cc:
```diff
@@ -0,0 +1,1125 @@
+                        CreateDataSetFromCSVData());
+
+  auto options = std::make_shared<arrow::dataset::ScanOptions>();
+  // sync scanning is not supported by ScanNode
+  options->use_async = true;
```
```suggestion
```

westonpace commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,451 @@ Datasets may be scanned multiple times; just make multiple scan
+``filter``
+----------
+
+``filter`` operation as the name suggests, provides an option to define a data filtering
+criteria. It keeps only rows matching a given expression. 
+Filters can be written using :class:`arrow::compute::Expression`. 
+For example, if we wish to keep rows of column ``b`` greater than 3, 
+then we can use the following expression::, can be written using 
+:class:`arrow::compute::FilterNodeOptions` as follows::
```
```suggestion
``filter`` operation, as the name suggests, provides an option to define data filtering
criteria. It selects rows matching a given expression. 
Filters can be written using :class:`arrow::compute::Expression`. 
For example, if we wish to keep rows where the value of column ``b``
is greater than 3,  then we can use the following expression:: 
```

Wording

vibhatha replied:
neat 👍

westonpace commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,451 @@ Datasets may be scanned multiple times; just make multiple scan
+                          //filter node options
+                          arrow::compute::FilterNodeOptions{filter_opt}));
+
+Filter Example;
```
```suggestion
Filter Example:
```

westonpace commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,451 @@ Datasets may be scanned multiple times; just make multiple scan
+``project`` operation rearranges, deletes, transforms, and creates columns.
+Each output column is computed by evaluating an expression
+against the source record batch. This is exposed via 
+:class:`arrow::compute::ProjectNodeOptions` class which requires, 
```
```suggestion
:class:`arrow::compute::ProjectNodeOptions` which requires, 
```

westonpace commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,451 @@ Datasets may be scanned multiple times; just make multiple scan
+Each output column is computed by evaluating an expression
+against the source record batch. This is exposed via 
+:class:`arrow::compute::ProjectNodeOptions` class which requires, 
+a :class:`arrow::compute::Expression`, names for the output columns (if names are not
+provided, the string representations of exprs will be used). 
```
```suggestion
an :class:`arrow::compute::Expression` and name for each of the output columns (if names are not
provided, the string representations of exprs will be used). 
```

westonpace commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,451 @@ Datasets may be scanned multiple times; just make multiple scan
+
+``aggregate`` operation provides various data aggregation options. 
+The :class:`arrow::compute::AggregateNodeOptions` is used to 
+define the aggregation criterion. These options can be 
+selected from :ref:`aggregation options <aggregation-option-list>`.
```
```suggestion
define the aggregation criteria. An aggregate node can first group data by
one or more key columns or the keys can be left off to compute aggregates
across the entire dataset.  Each aggregate node can compute any number of
aggregation functions.  Each aggregation function will be applied to every
field specified as a target.  The aggregation functions can be 
selected from :ref:`this list of aggregation functions <aggregation-option-list>`.

Note: This node is a "pipeline breaker" and will fully materialize the dataset in memory.
In the future, spillover mechanisms will be added which should alleviate this constraint.
```

Just adding a bit more info.

vibhatha replied:
This is very nice :) Updated the code.

westonpace commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,451 @@ Datasets may be scanned multiple times; just make multiple scan
+define the aggregation criterion. These options can be 
+selected from :ref:`aggregation options <aggregation-option-list>`.
+
+Example::
+
+  arrow::compute::IndexOptions index_options(arrow::MakeScalar("1"));
```
What is this for?

vibhatha replied:
typo from previously deleted options section.

westonpace commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,451 @@ Datasets may be scanned multiple times; just make multiple scan
+                            cp::MakeExecNode("aggregate", plan.get(), {source},
+                            aggregate_options));
+
+Aggregate example;
```
```suggestion
Aggregate example:
```

westonpace commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,451 @@ Datasets may be scanned multiple times; just make multiple scan
+                            aggregate_options));
+
+Aggregate example;
+
+Filter Example;
```
```suggestion
```

vibhatha replied:
typo.

westonpace commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,451 @@ Datasets may be scanned multiple times; just make multiple scan
+``sink``
+--------
+
+``sink`` operation can be considered as the option providing output or final node of an streaming 
+execution definition. :class:`arrow::compute::SinkNodeOptions` interface is used to pass 
+the required options. Requires 
+``arrow::AsyncGenerator<arrow::util::optional<cp::ExecBatch>>* generator``
+and ``arrow::util::BackpressureOptions backpressure``. 
+An execution plan should only have one "terminal" node (one sink node).
```
```suggestion
``sink`` operation can be considered as the option providing output or final node of an streaming 
execution definition. :class:`arrow::compute::SinkNodeOptions` interface is used to pass 
the required options. Requires 
``arrow::AsyncGenerator<arrow::util::optional<cp::ExecBatch>>* generator``
and ``arrow::util::BackpressureOptions backpressure``. 
An execution plan should only have one "terminal" node (one sink node).
```
```suggestion
``sink`` operation provides output and is the final node of a streaming 
execution definition. :class:`arrow::compute::SinkNodeOptions` interface is used to pass 
the required options. Similar to the source operator the sink operator exposes the output
with a function that returns a record batch future each time it is called.  It is expected the
caller will repeatedly call this function until the generator function is exhausted (returns
arrow::util::optional::nullopt).  If this function is not called often enough then record batches
will accumulate in memory.  An execution plan should only have one
"terminal" node (one sink node).  An execution plan may "finish" by marking
`exec_plan->finished()` as complete before the sink generator is fully consumed and the
execution plan can be safely destroyed without harming the sink generator (which will hold
references to the unconsumed batches).
```

Let's hide the details of backpressure for the moment.  I added a bit more information.

vibhatha replied:
Sure, I updated the section.

westonpace commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,451 @@ Datasets may be scanned multiple times; just make multiple scan
+``consuming_sink``
+------------------
+
+``consuming_sink`` operator is a sink operation containing consuming operation within the
+execution plan (i.e. the exec plan should not complete until the consumption has completed).
```
```suggestion
``consuming_sink`` operator is a sink operation containing consuming operation within the
execution plan (i.e. the exec plan should not complete until the consumption has completed).
Unlike the `sink` node this node takes in a callback function that is expected to consume the
batch.  Once this callback has finished the execution plan will no longer hold any reference to
the batch.

The consuming function may be called before a previous invocation has completed.  If the consuming
function does not run quickly enough then many concurrent executions could pile up, blocking the
CPU thread pool.  The execution plan will not be marked finished until all consuming function callbacks
have been completed.

Once all batches have been delivered the execution plan will wait for the `finish` future to complete
before marking the execution plan finished.  This allows for workflows where the consumption function
converts batches into async tasks (this is currently done internally for the dataset write node).
```

Some more detail, but I'm half tempted to hide this node.  Just like the `source` node this (and `sink`) are very tricky nodes to use correctly.

vibhatha replied:
Great, this is nice. Updated the docs.

westonpace commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,451 @@ Datasets may be scanned multiple times; just make multiple scan
+``order_by_sink``
+-----------------
+
+``order_by_sink`` operation is an extension to the ``sink`` operation. 
+This operation provides the ability to guarantee the ordering of the 
+stream by providing the :class:`arrow::compute::OrderBySinkNodeOptions`. 
+Here the :class:`arrow::compute::SortOptions` are provided to define which columns 
+are used for sorting and whether to sort by ascending or descending values.
```
```suggestion
``order_by_sink`` operation is an extension to the ``sink`` operation. 
This operation provides the ability to guarantee the ordering of the 
stream by providing the :class:`arrow::compute::OrderBySinkNodeOptions`. 
Here the :class:`arrow::compute::SortOptions` are provided to define which columns 
are used for sorting and whether to sort by ascending or descending values.

Note: This node is a "pipeline breaker" and will fully materialize the dataset in memory.
In the future, spillover mechanisms will be added which should alleviate this constraint.
```

westonpace commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,451 @@ Datasets may be scanned multiple times; just make multiple scan
+``select_k_sink`` option enables selecting k number of elements. 
+:class:`arrow::compute::SelectKOptions` which is a defined by 
+using :struct:`OrderBySinkNode` definition. This option returns a sink node that receives 
+inputs and then compute top_k/bottom_k.
```
```suggestion
inputs and then compute top_k/bottom_k.

Note: This node is a "pipeline breaker" and will fully materialize the input in memory.
In the future, spillover mechanisms will be added which should alleviate this constraint.
```

vibhatha replied:
updated

westonpace commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,451 @@ Datasets may be scanned multiple times; just make multiple scan
+``scan``
+---------
+
+`scan` is an operation used to load and process data, and the behavior of is defined using 
+:class:`arrow::dataset::ScanNodeOptions`. This option contains a set of definitions. 
+The :ref:`dataset<cpp-dataset-reading>` API also use scanner for processing data.
+In contrast to `source` operation, `scan` operation can load the data and apply scanning
+operations like filter and project to the loaded data. 
```
```suggestion
`scan` is an operation used to load and process datasets.  It should be preferred over the
more generic `source` node when your input is a dataset.  The behavior is defined using 
:class:`arrow::dataset::ScanNodeOptions`.  More information on datasets and the various
scan options can be found in :ref:`dataset<cpp-dataset-reading>`.

This node is capable of applying pushdown filters to the file readers which reduce
the amount of data that needs to be read.  This means you may supply the same
filter expression to the scan node that you also supply to the FilterNode because
the filtering is done in two different places.
```

vibhatha replied:
updated

westonpace commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,451 @@ Datasets may be scanned multiple times; just make multiple scan
+
+``write`` option enables writing a result to supported file formats (example `parquet`,
+`feather`, `csv`, etc). 
+The write options are provided via the :class:`arrow::dataset::WriteNodeOptions` and 
+defined using :class:`arrow::dataset::FileSystemDatasetWriteOptions`, 
+``std::shared_ptr<arrow::Schema>``, and 
+``std::shared_ptr<arrow::util::AsyncToggle> backpressure_toggle``. Here the 
+:class:`arrow::dataset::FileSystemDatasetWriteOptions` contains the meta-data required 
+to write the data. 
```
```suggestion
The write options are provided via the :class:`arrow::dataset::WriteNodeOptions` and 
defined using :class:`arrow::dataset::FileSystemDatasetWriteOptions`. Here the 
:class:`arrow::dataset::FileSystemDatasetWriteOptions` contains the meta-data required 
to write the data. 
```

Let's hide backpressure for now (that setting is optional I'm pretty sure).  It's actually a bug (my fault) that we require `schema`.  I think we can remove that in the future so let's skip mentioning it for the moment.

vibhatha replied:
Got it, updated the doc.

westonpace commented on cpp/examples/arrow/execution_plan_documentation_examples.cc:
```diff
@@ -0,0 +1,1125 @@
+                          MakeExecNode("source", plan.get(), {},
+                                       cp::SourceNodeOptions{input.schema, input.gen()}));
+  }
+  // TODO: decide whether to keep the filters or remove
```
Have you decided?

vibhatha replied:
Better to remove, we only need to show how a join works.

westonpace commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,451 @@ Datasets may be scanned multiple times; just make multiple scan
+Example List
+============
+
+There a set of examples can be found in ``examples/arrow/execution_plan_documentation_examples.cc``
```
```suggestion
There are examples of these nodes which can be found in ``examples/arrow/execution_plan_documentation_examples.cc``
```

westonpace commented on cpp/examples/arrow/execution_plan_documentation_examples.cc:
```diff
@@ -0,0 +1,1125 @@
+ * \param exec_context : execution context
+ * \return arrow::Status
+ */
+arrow::Status SourceHashJoinSinkExample(cp::ExecContext& exec_context) {
```
I don't see this example included in the rst docs.

vibhatha replied:
I missed this, thanks for noticing this one. Included the docs.

westonpace reviewed (changes requested):
This is a little overdue but I went through the examples in more detail and provided some extra information for the various operations.  This is a great set of examples!

vibhatha commented on cpp/examples/arrow/execution_plan_documentation_examples.cc:
```diff
@@ -0,0 +1,1125 @@
+
+// Demonstrate various operators in Arrow Streaming Execution Engine
+
+#define ABORT_ON_FAILURE(expr)                     \
```
Yes we can. Replaced and removed the macro.

vibhatha commented on cpp/examples/arrow/execution_plan_documentation_examples.cc:
```diff
@@ -0,0 +1,1125 @@
+    for (auto aug_name : {"__fragment_index", "__batch_index", "__last_in_fragment"}) {
+      names.emplace_back(aug_name);
+    }
+  }
```
Removed and updated.

vibhatha commented on cpp/examples/arrow/execution_plan_documentation_examples.cc:
```diff
@@ -0,0 +1,1125 @@
+
+  auto options = std::make_shared<arrow::dataset::ScanOptions>();
+  // sync scanning is not supported by ScanNode
+  options->use_async = true;
```
Rebased and updated the code.

vibhatha commented on cpp/examples/arrow/execution_plan_documentation_examples.cc:
```diff
@@ -0,0 +1,1125 @@
+          typename = typename std::enable_if<arrow::is_number_type<TYPE>::value |
+                                             arrow::is_boolean_type<TYPE>::value |
+                                             arrow::is_temporal_type<TYPE>::value>::type>
+arrow::Result<std::shared_ptr<arrow::Array>> GetArrayDataSample(
```
You're right, `ScanSinkExample` is in the middle of some util methods, by moving the example code, 
we can organize it better as you are suggesting. 

Moved the code and updated.

vibhatha commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,451 @@ Datasets may be scanned multiple times; just make multiple scan
+``source``
+----------
+
+`source` operation can be considered as an entry point to create a streaming execution plan. 
+A source node can be constructed as follows.
+:class:`arrow::compute::SourceNodeOptions` are used to create the ``source`` operation. 
+The :class:`Schema` of the data passing through and a function to generate data 
+``arrow::AsyncGenerator<arrow::util::optional<cp::ExecBatch>>``
+are required to create this option. Additionally, when using `source` operator, 
+the data scanning operations like filter and project may need to be applied
+in a later part of the execution plan. 
```
@westonpace 100% agree. For now let's update the documentation with this nice description, but later on I can update the documentation with new methods like "from_table".

vibhatha commented on cpp/examples/arrow/execution_plan_documentation_examples.cc:
```diff
@@ -0,0 +1,1125 @@
+  arrow::AsyncGenerator<arrow::util::optional<cp::ExecBatch>> sink_gen;
+
+  ARROW_ASSIGN_OR_RAISE(std::ignore, cp::MakeExecNode("sink", plan.get(), {scan},
+                                                      cp::SinkNodeOptions{&sink_gen}));
```
This is neat :)

vibhatha commented on cpp/examples/arrow/execution_plan_documentation_examples.cc:
```diff
@@ -0,0 +1,1125 @@
+
+  auto options = std::make_shared<arrow::dataset::ScanOptions>();
+  // sync scanning is not supported by ScanNode
+  options->use_async = true;
```
Updated and removed.

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,497 @@ Datasets may be scanned multiple times; just make multiple scan
+processing has begun.  This means we must supply the schema for a source node separately
+from the data itself.
+
+Struct to hold the data generator definition:
```
perhaps something like "We define a struct that provides everything we need: the in-memory batches, the schema, and a function that serves as the data generator."

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,497 @@ Datasets may be scanned multiple times; just make multiple scan
+  :linenos:
+  :lineno-match:
+
+Generating sample Batches for computation:
```
```suggestion
Generating sample batches for computation:
```

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,497 @@ Datasets may be scanned multiple times; just make multiple scan
+criteria. It selects rows matching a given expression. 
+Filters can be written using :class:`arrow::compute::Expression`. 
+For example, if we wish to keep rows where the value of column ``b``
+is greater than 3,  then we can use the following expression:: 
```
Rather a small nit, but can we consistently line-wrap these paragraphs? Text editors should have a shortcut to do this automatically.

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,497 @@ Datasets may be scanned multiple times; just make multiple scan
+
+``filter`` operation, as the name suggests, provides an option to define data filtering
+criteria. It selects rows matching a given expression. 
+Filters can be written using :class:`arrow::compute::Expression`. 
```
Hmm, `Expression` also doesn't get resolved for me. We should make sure to add anything that doesn't get resolved to the docs so they show up properly.

vibhatha replied:
Yes, same here it doesn't work for me. 
I included `arrow/compute/exec/expression.h` in `arrow/compute/api.h` as well. 
Is there an additional thing that we need to do to fix this? 
@westonpace what do you think?

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,497 @@ Datasets may be scanned multiple times; just make multiple scan
+arrow::util::optional::nullopt).  If this function is not called often enough then record batches
+will accumulate in memory.  An execution plan should only have one
+"terminal" node (one sink node).  An execution plan may "finish" by marking
+`exec_plan->finished()` as complete before the sink generator is fully consumed and the
```
Though, this is kind of an internal detail: maybe just say something like "An ExecPlan can terminate early due to cancellation or an error, before the output is fully consumed. However, the plan can be safely destroyed independently of the sink, which will hold the unconsumed batches."

vibhatha replied:
Rephrased and added the idea.

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,497 @@ Datasets may be scanned multiple times; just make multiple scan
+Constructing ``ExecNode`` using Options
+=======================================
+
+Using the execution plan we can construct various queries. 
+To construct such queries, we have provided a set of building blocks
+referred to as :class:`ExecNode` s. These nodes provide the ability to  
+construct operations like filtering, projection, join, etc. 
```
We should perhaps also link to the full example somehow for reference. (I'm not sure if we can get Sphinx to generate a page like it does for Python code, but maybe we can just include the full example at the bottom or on a separate page?)

vibhatha replied:
I included the full example at the bottom.

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,497 @@ Datasets may be scanned multiple times; just make multiple scan
+:class:`arrow::compute::SelectKOptions` which is a defined by 
+using :struct:`OrderBySinkNode` definition. This option returns a sink node that receives 
+inputs and then compute top_k/bottom_k.
+Note: This node is a "pipeline breaker" and will fully materialize the input in memory.
```
Use `.. note::` here too

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,497 @@ Datasets may be scanned multiple times; just make multiple scan
+``select_k_sink``
+-----------------
+
+``select_k_sink`` option enables selecting k number of elements. 
```
```suggestion
``select_k_sink`` option enables selecting the top/bottom K elements, similar to a SQL ``ORDER BY ... LIMIT K`` clause. 
```

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,497 @@ Datasets may be scanned multiple times; just make multiple scan
+Example List
+============
+
+There are examples of these nodes which can be found in ``examples/arrow/execution_plan_documentation_examples.cc``
```
Though, this list is pretty redundant with the list above, and as mentioned above, it might be nice to just include the example in its entirety.

vibhatha replied:
Yep, I just added it as an easy way to navigate through the examples, but since there is a menu in the right hand side, this is redundant in use and content-wise. Removing and settling for the simpler description suggested.

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,497 @@ Datasets may be scanned multiple times; just make multiple scan
+
+.. _stream_execution_hashjoin_docs:
+
+``hash_join``
```
nit, but maybe group `hash_join` with the non-sink nodes instead of putting it in the middle of the sink nodes here?

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,497 @@ Datasets may be scanned multiple times; just make multiple scan
+``hash_join`` operation provides the relational algebra operation, join using hash-based
+algorithm. :class:`arrow::compute::HashJoinNodeOptions` contains the options required in 
+defining a join. JoinType can be one of LEFT_SEMI, RIGHT_SEMI, LEFT_ANTI, RIGHT_ANTI,
+INNER, LEFT_OUTER, RIGHT_OUTER and FULL_OUTER. Also the join-key or by which column/s, 
```
Perhaps "hash_join supports left/right/full semi/anti/outerjoins"?

lidavidm replied:
Also "The join key (i.e. the column(s) to join on)" reads a bit more smoothly, I think

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,497 @@ Datasets may be scanned multiple times; just make multiple scan
+algorithm. :class:`arrow::compute::HashJoinNodeOptions` contains the options required in 
+defining a join. JoinType can be one of LEFT_SEMI, RIGHT_SEMI, LEFT_ANTI, RIGHT_ANTI,
+INNER, LEFT_OUTER, RIGHT_OUTER and FULL_OUTER. Also the join-key or by which column/s, 
+and output suffixes can be set via the the join options. 
```
What is an output suffix?

vibhatha replied:
Perhaps 

"The hash_join supports left/right/full semi/anti/outerjoins. 
Also the join-key (i.e. the column(s) to join on), and suffixes (i.e a suffix term like "_x"
which can be appended as a suffix for column names duplicated in both left and right 
relations.) can be set via the the join options." ?

lidavidm replied:
Thanks, I'd say we should explain that in a little more detail (or just omit it) since I don't think this is a standard feature of joins (at least in SQL, I may be wrong)

vibhatha replied:
Let me add a little more details.
Is it okay if we include external resources like peer reviewed papers, wiki pages, standard docs as helper materials?

lidavidm replied:
Yes, totally fine. (And I might just be ignorant of this if it's a common option.)

vibhatha replied:
Great, I will add more details to this.

vibhatha replied:
References and links added.

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,497 @@ Datasets may be scanned multiple times; just make multiple scan
+``write``
+---------
+
+``write`` option enables writing a result to supported file formats (example `parquet`,
+`feather`, `csv`, etc). 
+The write options are provided via the :class:`arrow::dataset::WriteNodeOptions` and 
+defined using :class:`arrow::dataset::FileSystemDatasetWriteOptions`. Here the 
+:class:`arrow::dataset::FileSystemDatasetWriteOptions` contains the meta-data required 
+to write the data. 
```
Perhaps

```
The ``write`` node saves query results as a dataset of files in a
format like Parquet, Feather, CSV, etc. using the :doc:`./dataset`
functionality in Arrow. The write options are provided via the
:class:`arrow::dataset::WriteNodeOptions` which in turn contains
:class:`arrow::dataset::FileSystemDatasetWriteOptions`.
:class:`arrow::dataset::FileSystemDatasetWriteOptions` provides
control over the written dataset, including options like the output
directory, file naming scheme, and so on.
```

vibhatha replied:
Neat 👍

lidavidm commented on cpp/examples/arrow/execution_plan_documentation_examples.cc:
```diff
@@ -0,0 +1,901 @@
+#include <utility>
+
+#include "arrow/array.h"
+#include "arrow/builder.h"
```
nit, but please be consistent with include style

lidavidm commented on cpp/examples/arrow/execution_plan_documentation_examples.cc:
```diff
@@ -0,0 +1,901 @@
+
+constexpr char kSep[] = "******";
+
+#define PRINT_BLOCK(msg)                                               \
+  std::cout << "" << std::endl;                                        \
+  std::cout << "\t" << kSep << " " << msg << " " << kSep << std::endl; \
+  std::cout << "" << std::endl;
```
I'm going to keep harping on this, but please at least define a function instead of a macro for this sort of thing.

lidavidm replied:
Also, this can all be one statement, or at the very least there is no need to print empty strings.

vibhatha replied:
Replacing with a simple function. 👍

lidavidm commented on cpp/examples/arrow/execution_plan_documentation_examples.cc:
```diff
@@ -0,0 +1,901 @@
+namespace cp = ::arrow::compute;
+
+std::string GetDataAsCsvString() {
+  std::string data_str = R"csv(a,b
```
Why not just constexpr this?

vibhatha replied:
Yes, it was replaced. No need to add a method.

lidavidm commented on cpp/examples/arrow/execution_plan_documentation_examples.cc:
```diff
@@ -0,0 +1,901 @@
+  arrow::io::IOContext io_context = arrow::io::default_io_context();
+  std::shared_ptr<arrow::io::InputStream> input;
+  std::string csv_data = GetDataAsCsvString();
+  arrow::util::string_view sv = csv_data;
```
Is it necessary to bounce through string_view here? Also, you might prefer `Buffer::FromString` to make it more obvious what's happening, and to avoid using Arrow's stdlib backports more than necessary.

vibhatha replied:
modified.

lidavidm commented on cpp/examples/arrow/execution_plan_documentation_examples.cc:
```diff
@@ -0,0 +1,901 @@
+  return record_batch->FromStructArray(struct_result);
+}
+
+arrow::Result<std::shared_ptr<arrow::dataset::Dataset>> CreateDataSetFromCSVData() {
```
I know I suggested using CSV originally, but if we're going to have both helpers like `GetArrayDataSample` and CSV, we should just pick one or the other and stick to it.

vibhatha replied:
Mmm... Let's pick the code instead of the CSV. It can provide more examples for the users in using Arrow APIs. 
I will work on a code block for this. 👍

vibhatha replied:
code added

vibhatha commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,497 @@ Datasets may be scanned multiple times; just make multiple scan
+   * - ``write``
+     - :class:`arrow::dataset::WriteNodeOptions`
+   * - ``union``
+     - N/A
```
the `WriteNodeOptions` works for me and that is included in `arrow/dataset/api.h`
@westonpace as we discussed before I included `arrow/compute/exec/options.h` in `arrow/compute/api.h`
But when I build the html, still it doesn't work (meaning: clickable link to class definition). I assume what @lidavidm referring is the same. @lidavidm please correct me if I am wrong.

vibhatha replied:
When building I did the `doxygen` thing at `cpp/apidocs` and then `make html` at `docs`.

westonpace replied:
Hmm...looks like it takes a bit more work to get things into the docs.  This (https://github.com/vibhatha/arrow/pull/1) should do just enough that we can get links.  The reference docs for expressions and exec plans are pretty rough at the moment but lets save cleaning that up for a future PR.

vibhatha replied:
Thank you @westonpace for this fix.

vibhatha commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,497 @@ Datasets may be scanned multiple times; just make multiple scan
+
+Using this option, the filter node can be constructed as follows::																
+
+  // creating filter node
+  arrow::compute::ExecNode* filter;
+    ARROW_ASSIGN_OR_RAISE(filter, arrow::compute::MakeExecNode("filter", 
+                          // plan
+                          plan.get(),
+                          // previous node
+                          {scan}, 
+                          //filter node options
+                          arrow::compute::FilterNodeOptions{filter_opt}));
```
I see your point. It is better to remove those comments.

vibhatha commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,497 @@ Datasets may be scanned multiple times; just make multiple scan
+          // previous node 
+          {scan},
+          // project node options 
+          arrow::compute::ProjectNodeOptions{{a_times_2}}));
```
I removed the comments from here. It is already used for an example for projection. 
For source, I just kept a very simple example.

vibhatha commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,497 @@ Datasets may be scanned multiple times; just make multiple scan
+  :start-after: (Doc section: Project Example)
+  :end-before: (Doc section: Project Example)
+  :linenos:
+  :lineno-match:
```
included as a comment within the code.

vibhatha commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,497 @@ Datasets may be scanned multiple times; just make multiple scan
+``aggregate``
+-------------
+
+``aggregate`` operation provides various data aggregation options. 
+The :class:`arrow::compute::AggregateNodeOptions` is used to 
+define the aggregation criteria. An aggregate node can first group data by
+one or more key columns or the keys can be left off to compute aggregates
+across the entire dataset.  Each aggregate node can compute any number of
+aggregation functions.  Each aggregation function will be applied to every
+field specified as a target.  The aggregation functions can be 
+selected from :ref:`this list of aggregation functions <aggregation-option-list>`.
+Note: This node is a "pipeline breaker" and will fully materialize the dataset in memory.
+In the future, spillover mechanisms will be added which should alleviate this constraint.
```
Nice description 👍

vibhatha commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,497 @@ Datasets may be scanned multiple times; just make multiple scan
+  :start-after: (Doc section: Aggregate Example)
+  :end-before: (Doc section: Aggregate Example)
+  :linenos:
+  :lineno-match:
```
I will work on additional example 👍

vibhatha replied:
additional example added

vibhatha commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,497 @@ Datasets may be scanned multiple times; just make multiple scan
+  arrow::compute::ExecNode* sink;
+
+  ARROW_ASSIGN_OR_RAISE(sink, arrow::compute::MakeExecNode("sink", plan.get(), {source},
+                                                arrow::compute::SinkNodeOptions{&sink_gen}));
```
Yeah, I understand what you are pointing out. May be since we add the excerpts, we can omit these.

vibhatha replied:
I removed similar excerpts like that.

vibhatha commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,497 @@ Datasets may be scanned multiple times; just make multiple scan
+stream by providing the :class:`arrow::compute::OrderBySinkNodeOptions`. 
+Here the :class:`arrow::compute::SortOptions` are provided to define which columns 
+are used for sorting and whether to sort by ascending or descending values.
+Note: This node is a "pipeline breaker" and will fully materialize the dataset in memory.
```
removed similar ones.

vibhatha commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,497 @@ Datasets may be scanned multiple times; just make multiple scan
+`scan` is an operation used to load and process datasets.  It should be preferred over the
+more generic `source` node when your input is a dataset.  The behavior is defined using 
+:class:`arrow::dataset::ScanNodeOptions`.  More information on datasets and the various
+scan options can be found in :ref:`dataset<cpp-dataset-reading>`.
```
Make sense, it's neater. Haven't used it before. But suits well here. 👍 
Added it.

westonpace reviewed (approved):
I think we are approaching a state where this is a clear improvement over what we had before (i.e. nothing) and any major changes we should probably defer to future PRs.  So let's just focus comments at this point on cleaning up this PR.

vibhatha commented on cpp/examples/arrow/execution_plan_documentation_examples.cc:
```diff
@@ -0,0 +1,906 @@
+}
+
+// (Doc section: Materialize)
+cp::Expression Materialize(std::vector<std::string> names) {
+  std::vector<cp::Expression> exprs;
+  for (const auto& name : names) {
+    exprs.push_back(cp::field_ref(name));
+  }
+
+  return cp::project(exprs, names);
```
Of course, I have missed this. I will update it.

vibhatha commented:
@westonpace I will clean up the PR and make it ready. 
Thanks a lot for taking time and working on the reviews.

lidavidm commented on docs/source/cpp/streaming_execution.rst:
```diff
@@ -305,3 +305,427 @@ Datasets may be scanned multiple times; just make multiple scan
+``source``
+----------
+
+A `source` operation can be considered as an entry point to create a streaming execution plan. 
```
```suggestion
A ``source`` operation can be considered as an entry point to create a streaming execution plan. 
```

lidavidm replied:
There's some other instances of this below

lidavidm commented on cpp/examples/arrow/execution_plan_documentation_examples.cc:
```diff
@@ -0,0 +1,925 @@
+ */
+arrow::Result<std::shared_ptr<arrow::dataset::Dataset>> GetDataset() {
+  auto null_long = std::numeric_limits<int64_t>::quiet_NaN();
+  auto null_bool = std::numeric_limits<bool>::quiet_NaN();
```
Is this valid? `bool` has no NaN value.

lidavidm replied:
We probably need a separate overload that takes both an array of values and an array of validity bits.

lidavidm commented on cpp/examples/arrow/execution_plan_documentation_examples.cc:
```diff
@@ -0,0 +1,925 @@
+  std::shared_ptr<arrow::RecordBatch> record_batch;
+  ARROW_ASSIGN_OR_RAISE(auto res_batch, GetSampleRecordBatch(array_vector, field_vector));
+  cp::ExecBatch batch{*res_batch};
+  arrow::Result<cp::ExecBatch> result(batch);
+  return result;
```
Result has an implicit constructor, just return `batch`

lidavidm replied:
And ditto below since there's a few other occurences of this

lidavidm commented on cpp/examples/arrow/execution_plan_documentation_examples.cc:
```diff
@@ -0,0 +1,925 @@
+  // // stop producing
+  plan->StopProducing();
+  // // plan mark finished
+  auto future = plan->finished();
+  ARROW_RETURN_NOT_OK(future.status());
+  future.Wait();
+  return arrow::Status::OK();
```
It doesn't make sense to `Wait` after `status()` since `status()` has to wait on the future. Just `return future.status()`?

lidavidm commented on cpp/examples/arrow/execution_plan_documentation_examples.cc:
```diff
@@ -0,0 +1,925 @@
+                        MakeExecNode("consuming_sink", plan.get(), {source},
+                                     cp::ConsumingSinkNodeOptions(consumer)));
+
+  ABORT_NOT_OK(consuming_sink->Validate());
```
Use ARROW_RETURN_NOT_OK

lidavidm replied:
And the same below.

lidavidm commented on cpp/examples/arrow/execution_plan_documentation_examples.cc:
```diff
@@ -0,0 +1,925 @@
+  std::cout << "Source Finished!" << std::endl;
+  // Mark consumption complete, plan should finish
+  arrow::Status finish_status;
+  // finish.Wait();
```
Remove any commented code

lidavidm commented on cpp/examples/arrow/execution_plan_documentation_examples.cc:
```diff
@@ -0,0 +1,925 @@
+  // finish.Wait();
+  finish.MarkFinished(finish_status);
+  ARROW_RETURN_NOT_OK(plan->finished().status());
+  ARROW_RETURN_NOT_OK(finish_status);
```
This line is redundant, finish_status is OK

lidavidm replied:
And actually, `finish.MarkFinished(Status::OK());` above is probably clearer since it's not necessarily obvious that a default-constructed Status is OK

lidavidm commented on cpp/examples/arrow/execution_plan_documentation_examples.cc:
```diff
@@ -0,0 +1,925 @@
+
+  ARROW_RETURN_NOT_OK(cp::MakeExecNode("write", plan.get(), {scan}, write_node_options));
+
+  ABORT_NOT_OK(plan->Validate());
```
Use ARROW_RETURN_NOT_OK

lidavidm reviewed (commented):
Thanks, just some consistency things in the code.

vibhatha commented:
@lidavidm added some modifications based on the review comments.

lidavidm reviewed (approved):
Thanks for the updates.

westonpace commented:
This appveyor CI failure looks legitimate:

```
FAILED: examples/arrow/CMakeFiles/execution_plan_documentation_examples.dir/Unity/unity_0_cxx.cxx.obj 
C:\Miniconda37-x64\Scripts\clcache.exe  /nologo /TP -DARROW_HAVE_RUNTIME_AVX2 -DARROW_HAVE_RUNTIME_AVX512 -DARROW_HAVE_RUNTIME_BMI2 -DARROW_HAVE_RUNTIME_SSE4_2 -DARROW_HAVE_SSE4_2 -DARROW_MIMALLOC -DARROW_WITH_RE2 -DARROW_WITH_UTF8PROC -DAWS_AUTH_USE_IMPORT_EXPORT -DAWS_CAL_USE_IMPORT_EXPORT -DAWS_CHECKSUMS_USE_IMPORT_EXPORT -DAWS_COMMON_USE_IMPORT_EXPORT -DAWS_COMPRESSION_USE_IMPORT_EXPORT -DAWS_CRT_CPP_USE_IMPORT_EXPORT -DAWS_EVENT_STREAM_USE_IMPORT_EXPORT -DAWS_HTTP_USE_IMPORT_EXPORT -DAWS_IO_USE_IMPORT_EXPORT -DAWS_MQTT_USE_IMPORT_EXPORT -DAWS_MQTT_WITH_WEBSOCKETS -DAWS_S3_USE_IMPORT_EXPORT -DAWS_SDKUTILS_USE_IMPORT_EXPORT -DAWS_SDK_VERSION_MAJOR=1 -DAWS_SDK_VERSION_MINOR=9 -DAWS_SDK_VERSION_PATCH=178 -DAWS_USE_IO_COMPLETION_PORTS -DBOOST_ALL_DYN_LINK -DBOOST_ALL_NO_LIB -DGTEST_LINKED_AS_SHARED_LIBRARY=1 -DPROTOBUF_USE_DLLS -DUSE_IMPORT_EXPORT -DUSE_IMPORT_EXPORT=1 -DUSE_WINDOWS_DLL_SEMANTICS -D_CRT_SECURE_NO_WARNINGS -D_ENABLE_EXTENDED_ALIGNED_STORAGE -Isrc -I..\src -I..\src\generated -I..\thirdparty\flatbuffers\include -IC:\Miniconda37-x64\envs\arrow\Library\include -Imimalloc_ep\src\mimalloc_ep\include\mimalloc-1.7 -Igoogletest_ep-prefix\include -Ixsimd_ep\src\xsimd_ep-install\include -I..\thirdparty\hadoop\include /DWIN32 /D_WINDOWS  /GR /EHsc /D_SILENCE_TR1_NAMESPACE_DEPRECATION_WARNING   /EHsc /wd5105 /bigobj /utf-8 /W3 /wd4800 /wd4996 /wd4065  /WX /MP /MD /Od /UNDEBUG /showIncludes /Foexamples\arrow\CMakeFiles\execution_plan_documentation_examples.dir\Unity\unity_0_cxx.cxx.obj /Fdexamples\arrow\CMakeFiles\execution_plan_documentation_examples.dir\ /FS -c examples\arrow\CMakeFiles\execution_plan_documentation_examples.dir\Unity\unity_0_cxx.cxx
..\src\arrow/testing/gtest_util.h(31): fatal error C1083: Cannot open include file: 'gtest/gtest.h': No such file or directory
[83/314] Building CXX object examples\arrow\CMakeFiles\flight_grpc_example.dir\Unity\unity_0_cxx.cxx.obj
[84/314] Linking CXX shared library release\arrow.dll
   Creating library release\arrow.lib and object release\arrow.exp
ninja: build stopped: subcommand failed.
```

I think the problem is that `execution_plan_documentation_examples.h` is including gtest via the `arrow/compute/exec/test_util.h` import.  Luckily I don't think we are using that import for much.  Can you try removing that and see if the CI build passes.

westonpace commented:
Thank you @vibhatha , this was quite the undertaking.

vibhatha commented:
Thank you for the support @westonpace @lidavidm  👍
